### PR TITLE
make AQL modification operations async

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Forward-port of #14525, see there for details.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/Aql/SimpleModifier.cpp [hunk 1]
```diff
@@ -198,12 +303,35 @@ bool SimpleModifier<ModifierCompletion, Enable>::resultAvailable() const {
...
+                        [](std::exception_ptr const&) { return true; },
+                    },
+                    _results);
+}
+
+template <typename ModifierCompletion, typename Enable>
+bool SimpleModifier<ModifierCompletion, Enable>::hasNoResultOrOperationPending() const noexcept {
+  return std::visit(overload{
+                        [](NoResult) { return true; },
+                        [](Waiting) { return false; },
```
Comment (jsteemann) [inline arangod/Aql/SimpleModifier.cpp:328]: Given the method name, shouldn't this case return `true` as well?

File: arangod/Aql/SimpleModifier.h [hunk 2]
```diff
@@ -71,11 +74,16 @@ struct is_modifier_completion_trait<UpdateReplaceModifierCompletion> : std::true
 };
 
 template <typename ModifierCompletion, typename Enable = typename std::enable_if_t<is_modifier_completion_trait<ModifierCompletion>::value>>
-class SimpleModifier {
+class SimpleModifier : public std::enable_shared_from_this<SimpleModifier<ModifierCompletion, Enable>> {
+  /// @brief mutex that pr
```
Comment (jsteemann) [inline arangod/Aql/SimpleModifier.h:78]: comment is cut off

File: arangod/Aql/SimpleModifier.cpp (hunk 1 above)
Comment (goedderz) [reply arangod/Aql/SimpleModifier.cpp:328]: The name is chosen unwisely. What I meant was, there is neither a result, nor an operation pending. So `NoResult` is the only true case. Perhaps `hasNeitherResultNorOperationPending` would be less ambiguous.

Comment (goedderz) [reply arangod/Aql/SimpleModifier.cpp:328]: Maybe there's also a better name for `NoResult`; e.g. `hasNoResult` could mean both having `NoResult`, or not having `OperationResult`. Perhaps `Nothing` or `Empty`?

Comment (jsteemann) [reply arangod/Aql/SimpleModifier.cpp:328]: It is fine then. No need to change.

File: arangod/Aql/ModificationExecutorAccumulator.h [hunk 3]
```diff
@@ -58,12 +60,10 @@ class ModificationExecutorAccumulator {
     _accumulator.openArray();
   }
 
-  size_t nrOfDocuments() const { return _accumulator.slice().length(); }
+  [[nodiscard]] size_t nrOfDocuments() const { return _accumulator.slice().length(); }
```
Comment (mchacki) [inline arangod/Aql/ModificationExecutorAccumulator.h:63]: I think this is correct.
But shall we add an assertion here that _accumulator is a closed array whenever this method is called?
For instance this is broken as long as we are still accumulating things, as we will not be able to get the Slice, which will throw a "builder not yet sealed" which is hard to find.

File: arangod/Aql/SimpleModifier.cpp (hunk 1 above)
Comment (mchacki) [reply arangod/Aql/SimpleModifier.cpp:328]: I also it confused by this methods name.
maybe: `hasNothingInFlight` or `hasNothingInProcess` ?
As far as I understood the NoResult indicates: I am ready for more work and not waiting on a response, or someone handling it.
The name `NoResult` and the camel case of `no result` => `NoResult` is causing the ambiguity here.
I actually think the NoResult type is good.

File: arangod/Aql/SimpleModifier.h [hunk 4]
```diff
@@ -103,56 +111,60 @@ class SimpleModifier {
...
   std::vector<ModOp> _operations;
   ModificationExecutorAccumulator _accumulator;
 
-  OperationResult _results;
-
-  std::vector<ModOp>::const_iterator _operationsIterator;
-  VPackArrayIterator _resultsIterator;
-
   size_t const _batchSize;
+
```
Comment (mchacki) [inline arangod/Aql/SimpleModifier.h:161]: As @jsteemann mentioned above:
```suggestion
  ///@brief mutex that protects the _results.
```

**Review by mchacki — APPROVED** [review]
LGTM.

I have left some minor suggestions, nothing I would block this PR for.

File: arangod/Aql/SimpleModifier.cpp (hunk 1 above)
Comment (goedderz) [reply arangod/Aql/SimpleModifier.cpp:328]: I changed the name.

File: arangod/Aql/SimpleModifier.h (hunk 2 above)
Comment (goedderz) [reply arangod/Aql/SimpleModifier.h]: done

File: arangod/Aql/ModificationExecutorAccumulator.h (hunk 3 above)
Comment (goedderz) [reply arangod/Aql/ModificationExecutorAccumulator.h]: added an assertion

File: arangod/Aql/SimpleModifier.h (hunk 4 above)
Comment (goedderz) [reply arangod/Aql/SimpleModifier.h:160]: added the comment here

**Pull request merged**